src/main/java/frc/robot/systems/DriveFSMSystem.java

```diff
@@ -0,0 +1,410 @@
+        double error = degrees - getHeading();
+        if (error <= Constants.TURN_ERROR_THRESHOLD_DEGREE) {
+            finishedTurning = true;
+            return;
+        }
+        double power = error / Constants.TURN_ERROR_POWER_RATIO;
+        if (Math.abs(power) < Constants.MIN_TURN_POWER) {
+            power = Constants.MIN_TURN_POWER * power < 0 ? -1 : 1;
+        }
+
+        frontLeftMotor.set(power);
+        frontRightMotor.set(-power);
+        backLeftMotor.set(power);
+        backRightMotor.set(-power);
+    }
+
+    /**
+    * Gets the heading from the gyro.
+    * @return the gyro heading
+    */
+    private double getHeading() {
+        return 90 - imu.getGyroAngleZ();
+    }
+
+    private void handleTeleOpState(TeleopInput input) {
+        if (input == null) {
+            return;
+        }
+
+        double joystickY = input.getDrivingJoystickY();
+        double steerAngle = input.getSteerAngle();
+        double currentLeftPower = frontLeftMotor.get();
+        double currentRightPower = frontRightMotor.get();
+
+        if(input.isForwardDrivingButtonPressed()) {
+            isDrivingForward = true;
+        }else if(input.isBackwardDrivingButtonPressed()) {
+            isDrivingForward = false;
+        }
+
+        DrivePower targetPower = DriveModes.arcadedrive(joystickY, steerAngle, currentLeftPower, 
+        currentRightPower, isDrivingForward);
+
+        //multiple speed modes
+        if (input.getTriggerPressed()) {
+            targetPower.scale(Constants.MAX_POWER);
+        } else {
+            targetPower.scale(Constants.REDUCED_MAX_POWER);
+        }
+
+        DrivePower power;
+
+        //acceleration
+        power = Functions.accelerate(targetPower, new DrivePower(currentLeftPower, currentRightPower));
+
+        //turning in place
+        if (Math.abs(joystickY) < Constants.TELEOP_MIN_MOVE_POWER) {
+            power = Functions.turnInPlace(joystickY, steerAngle);
+        }
+
+        leftPower = power.getLeftPower();
+        rightPower = power.getRightPower();
+
+        System.out.println("Ecoder left: " + frontLeftMotor.getEncoder().getPosition());
+        System.out.println("Encoder right: " + frontRightMotor.getEncoder().getPosition());
+
+        frontRightMotor.set(rightPower);
+        frontLeftMotor.set(leftPower);
+        backRightMotor.set(rightPower);
+        backLeftMotor.set(leftPower);
+
+        //check if voltage can be negative
+        // frontRightMotor.setVoltage(rightPower * 12);
+        // frontLeftMotor.setVoltage(leftPower * 12);
+        // backRightMotor.setVoltage(rightPower * 12);
+        // backLeftMotor.setVoltage(leftPower * 12);
+        
+    }
+
+    private double limitPower(double number) {
+        if (number > 1) {
+            return 1;
+        }
+        if (number < -1) {
+            return -1;
+        }
+        return number;
+    }
+
+    private void updateLineOdometry() {
```

Comment:
I didn't hear anything about odometry during your weekly update, how's it going?

Reply:
We're testing it. We had to find an encoder constant to multiply to the obtained encoder values since inaccuracies (such as deflated pneumatic tires) caused the encoder values to not reflect the actual distance traveled. But now that we found that constant, the odometry calculations seem to be working, but it still needs further testing.